### PR TITLE
Wait until body file descriptor is open

### LINKED ISSUE #1046: Body stream error is not caught by Got
#### Describe the bug

 - Node.js version: 12.14.1
 - OS & version: MacOS 10.13.4

When using streams to upload a file, an error thrown from the stream read (for example, passing an inexistent file path) is not being caught by `got`.

This looks similar to #614

#### Actual behavior
The code in the example throws an error that is not caught by `got` itself.
Console output:
```
events.js:200
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open './no-file.txt'
Emitted 'error' event on ReadStream instance at:
    at internal/fs/streams.js:120:12
    at FSReqCallback.oncomplete (fs.js:146:23) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: './no-file.txt'
}
```

#### Expected behavior
The error should be caught and "caught error" should be logged to the console.

#### Code to reproduce

```js
const got = require('got')
const fs = require('fs')

got
  .put('http://a.com/', {
    body: fs.createReadStream('./file-that-does-not-exist.txt'),
  })
  .catch(() => console.log('caught error'))
```

PR with failing test: https://github.com/sindresorhus/got/pull/1047

#### Checklist

- [x] I have read the documentation.
- [x] I have tried my code with the latest version of Node.js and Got.

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes #1046

### CODE REVIEW

File: source/create.ts [hunk 1]
```diff
@@ -99,7 +100,29 @@ const aliases: readonly HTTPAlias[] = [
...
+				(result as ProxyStream).destroy(error);
+				return;
+			}
+
+			(result as CancelableRequest<unknown>).cancel(error.message);
+		};
+
+		options.body.once('error', errorHandler);
+
+		// @ts-ignore Each member of the union type '...' has signatures, but none of those signatures are compatible with each other
```
Comment (szmarczak) [inline source/create.ts]: I think it just says that result can be either or a stream or a promise and it doesn't know what types to choose

**Comment by szmarczak** [comment]
Or it could be done via a promise that waits until `options.body` is writable...

File: source/request-as-event-emitter.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import {ReadStream} from 'fs';
```
Comment (sindresorhus) [inline source/request-as-event-emitter.ts]: Generally, it's nicer to rename generic imports like this as it's not immediately clear further down in the code that it's an **fs** read stream.

```
import {ReadStream: FsReadStream} from 'fs';
```

Comment (szmarczak) [reply source/request-as-event-emitter.ts]: Will do :)

**Pull request merged**